snafu/benchmarks/nighthawk/nighthawk.py

```diff
@@ -0,0 +1,276 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Wrapper for running the nighthawk workload. See https://github.com/envoyproxy/nighthawk for more information."""
+import dataclasses
+import datetime
+import re
+import shlex
+import sys
+import json
+import os
+import socket
+import numpy as np
+import subprocess
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+from snafu.benchmarks import Benchmark, BenchmarkResult
+from snafu.config import Config, ConfigArgument, FuncAction, check_file, none_or_type
+
+
+@dataclasses.dataclass
+class NighthawkStat:
+    """Parsed Nighthawk Statistic."""
+
+    workload: str
+    uuid: str
+    user: str
+    cluster_name: str
+    duration: int
+    targets: List[str]
+    concurrency: int
+    connections: int
+    max_requests_per_connection: int
+    hostname: str
+    requested_qps: float
+    throughput: float
+    status_codes_1xx: float
+    status_codes_2xx: float
+    status_codes_3xx: float
+    status_codes_4xx: float
+    status_codes_5xx: float
+    p50_latency: float
+    p75_latency: float
+    p80_latency: float
+    p90_latency: float
+    p95_latency: float
+    p99_latency: float
+    p99_9_latency: float
+    avg_latency: float
+    timestamp: str
+    bytes_in: float
+    bytes_out: float
+    iteration: Optional[int] = None
+
+
+@dataclasses.dataclass
+class NighthawkConfig:
+    """Container for common configuration options that are passed to Nighthawk."""
+
+    concurrency: Optional[int] = None
+    duration: Optional[int] = None
+    connections: Optional[int] = None
+    max_requests_per_connection: Optional[int] = None
+    rps: Optional[int] = None
+    kind: Optional[str] = None
+    url: Optional[str] = None
+
+    @classmethod
+    def new(cls, stdout: NighthawkStat, config: Config):
+        """Create a new instance given instances of :py:mod:`~snafu.config.Config` and NighthawkStat."""
+
+        kwargs: Dict[str, Any] = {}
+        for fields in dataclasses.fields(cls):
+            val = getattr(stdout, fields.name, None)
+            if val is None:
+                val = getattr(config, fields.name, None)
+            kwargs[fields.name] = val
+        return cls(**kwargs)
+
+
+class Nighthawk(Benchmark):
+    """Wrapper for the nighthawk benchmark."""
+
+    tool_name = "nighthawk"
+    args = (
+        ConfigArgument(
+            "-s",
+            "--sample",
```

Comment:
Nit but this should be pluralized to samples

Reply:
Done